### PR TITLE
Add Cisco Meraki cloud test

### PR DESCRIPTION
**Pull request by jiribrejcha**
## Type of change 

*Pick at least one.*

* [ ] Bug fix (non-breaking change that fixes something)
* [ ] Documentation update (readme, changelog, man page, etc.)
* [x] New feature (non-breaking change adding functionality)
* [ ] Breaking change (fix or feature changing existing functionality)
* [ ] Code quality improvement (refactor, performance improvements)
* [ ] Other (please specify):

## Breaking change

*Does this Pull Request introduce a breaking change?* 
No.

- What functionality breaks?
- Why does it break?
- How should it be migrated?  
- Are there any backward-compatible alternatives?

## Proposed change

*Please describe the purpose of this change, the problem it solves, and why the change is necessary. Including code snippets or links to related documentation when applicable will assist approval.*

Adds Cisco Meraki cloud test to FPMS.

## Testing

*Please explain how you tested this change manually, and if applicable, what new tests you added.*

- [ ] Did you add new automated tests? If yes, explain which edge cases are covered.
- [ ] Does this change affect any existing tests? If so, how did you adjust them?
- [ ] Please provide test run results or screenshots if applicable.

## Checklist

* [x] I have read the [contribution guidelines and policies](https://github.com/WLAN-Pi/.github/blob/main/docs/contributing.md)
* [x] I have targeted this PR against the correct git branch
* [x] I ran the test suite and verified it succeeded
* [x] I linked an approved GitHub issue to this PR (in the next section). No PR without a related GH issue. Conversations about your PR efforts in other channels such as electronic mail, social media, morse code, and WLAN Pi Slack **do not count**.
* [x] I have updated the changelog (if applicable)
* [ ] I have added or updated the documentation (if applicable)

## Related Issues/PRs

*Pick at least one.*

- This PR fixes/closes issue #132 
- This PR is related to issue #
- This PR depends on/blocks PR #

### CODE REVIEW

File: fpms/modules/cloud_tests.py [hunk 1]
```diff
@@ -159,6 +160,198 @@ def test_aruba_cloud(self, g_vars):
...
+        """
+
+        def get_default_gateway():
+            try:
+                result = subprocess.check_output("ip route | awk '/default/ {print $3}'", shell=True, text=True).strip()
+                return result
+            except subprocess.CalledProcessError:
+                return None
+
+        def test_udp(ip, port, timeout=15):
```
Comment (joshschmelzle) [inline fpms/modules/cloud_tests.py]: This function isn't being used. Delete it.

File: fpms/modules/cloud_tests.py [hunk 2]
```diff
@@ -159,6 +160,198 @@ def test_aruba_cloud(self, g_vars):
...
+        UDP port 7351 is only used by Meraki devices running older firmware and we can't reliably test UDP anyway. Skipping this test.
+
+        APs perform connnection tests:
+        - ping 8.8.8.8
+        - ARP gateway - skipping, if we deal with gateway issues, most other tests will fail
+        - DNS resolution
+
+        """
+
+        def get_default_gateway():
```
Comment (joshschmelzle) [inline fpms/modules/cloud_tests.py]: This function is called exactly once. Since the scope of this function is inside test_meraki_cloud() it can't be reused anywhere else. Probably not needed to have a function for this.

File: fpms/modules/cloud_tests.py [hunk 3]
```diff
@@ -159,6 +160,198 @@ def test_aruba_cloud(self, g_vars):
...
+
+        def test_ping(host, timeout=2):
+            try:
+                subprocess.check_call(f"ping -c1 -W{timeout} -4 -q {host}", shell=True, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
+                return True
+            except subprocess.CalledProcessError:
+                return False
+
+        # Set a global timeout for socket operations
+        socket.setdefaulttimeout(2)
```
Comment (joshschmelzle) [inline fpms/modules/cloud_tests.py]: Remove this line. This impacts global settings that can impact behavior for other tests. Set the timeout at the socket instance level.

```
s = socket.socket(...)
s.settimeout(2)
r = s.connect_ex(('host', port))
```

File: fpms/modules/cloud_tests.py [hunk 4]
```diff
@@ -159,6 +160,198 @@ def test_aruba_cloud(self, g_vars):
...
+                if data:
+                    return True
+            except (socket.timeout, Exception):
+                return False
+            finally:
+                client.close()
+
+        def test_dns(hostname):
+            try:
+                # Dig with timeout is faster to fail and more flexible than native Python
```
Comment (joshschmelzle) [inline fpms/modules/cloud_tests.py:223]: Are we switching DNS testing in this module to dig vs socket.gethostbyname() just for faster failures due to timeout? 

If there are other reasons besides adjusting the timeout, it might make sense to change the scope of this function so that other tests can later be refactored (with a different PR) and reuse this code.

File: fpms/modules/cloud_tests.py [hunk 5]
```diff
@@ -1,5 +1,6 @@
 import subprocess
 import socket
+import struct
```
Comment (joshschmelzle) [inline fpms/modules/cloud_tests.py]: This import isn't used anywhere. Delete it.

File: fpms/modules/cloud_tests.py [hunk 6]
```diff
@@ -159,6 +160,198 @@ def test_aruba_cloud(self, g_vars):
...
+        # Ignore any more key presses as this could cause us issues
+        g_vars["disable_keys"] = True
+
+        # Has test been run already?
+        if g_vars["result_cache"] == False:
+
+            # Record test success/fail
+            test_fail = False
+
+            gateway_ip = get_default_gateway()
```
Comment (joshschmelzle) [inline fpms/modules/cloud_tests.py]: This code does nothing with the `gateway_ip` variable. Is it needed for a test? If not, delete this line and the unused get_default_gateway func.

**Review by joshschmelzle — DISMISSED** [review]
Left a few comments that suggest changes which improve the code.

File: fpms/modules/cloud_tests.py (hunk 3 above)
Comment (joshschmelzle) [reply fpms/modules/cloud_tests.py]: Do not modify the global socket setting in this location. This might cause unexpected side effects in other parts of the codebase that use sockets. If we need to set a global setting, we should do it in a central location in the codebase.

File: fpms/modules/cloud_tests.py (hunk 4 above)
Comment (jiribrejcha) [reply fpms/modules/cloud_tests.py:223]: Yes, dig fails significantly faster when UDP 53 port is blocked upstream. Which improves the UI - we get output on display faster.

Shortening the socket.gethostbyname() timeout does not improve this scenario. It waits until name resolution fails, and then eventually gives up. I haven't found a way to speed that up natively in Python.

File: fpms/modules/cloud_tests.py (hunk 3 above)
Comment (jiribrejcha) [reply fpms/modules/cloud_tests.py]: Done

File: fpms/modules/cloud_tests.py (hunk 1 above)
Comment (jiribrejcha) [reply fpms/modules/cloud_tests.py]: Done. Thank you.

File: fpms/modules/cloud_tests.py (hunk 2 above)
Comment (jiribrejcha) [reply fpms/modules/cloud_tests.py]: Removed.

File: fpms/modules/cloud_tests.py (hunk 5 above)
Comment (jiribrejcha) [reply fpms/modules/cloud_tests.py]: Deleted.

File: fpms/modules/cloud_tests.py (hunk 6 above)
Comment (jiribrejcha) [reply fpms/modules/cloud_tests.py]: Deleted.

File: fpms/modules/cloud_tests.py (hunk 4 above)
Comment (joshschmelzle) [reply fpms/modules/cloud_tests.py:223]: Ok, we can reuse this code for other tests at some point then.

**Review by joshschmelzle — APPROVED** [review]
Thanks for making changes to clean this up a bit.

**Comment by joshschmelzle** [comment]
Closes #133
